oracc-hyperestraier/man/estcmd.1
.TH ESTCMD 1 "2007-03-06" "Man Page" "Hyper Estraier"

.SH NAME
estcmd \- command line interface of the core API

.SH SYNOPSIS
.PP
.B estcmd create [\-tr] [\-apn|\-acc] [\-xs|\-xl|\-xh|\-xh2|\-xh3] [\-sv|\-si|\-sa] [\-attr name type] db
.PP
.B estcmd put [\-tr] [\-cl] [\-ws] [\-apn|\-acc] [\-xs|\-xl|\-xh||\-xh2|\-xh3] [\-sv|\-si|\-sa] db [file]
.PP
.B estcmd out [\-cl] [\-pc enc] db expr
.PP
.B estcmd edit [\-pc enc] db expr name [value]
.PP
.B estcmd get [\-nl|\-nb] [\-pidx path] [\-pc enc] db expr [attr]
.PP
.B estcmd list [\-nl|\-nb] [\-lp] db
.PP
.B estcmd uriid [\-nl|\-nb] [\-pidx path] [\-pc enc] db expr
.PP
.B estcmd meta db [name [value]]
.PP
.B estcmd inform [\-nl|\-nb] db
.PP
.B estcmd optimize [\-onp] [\-ond] db
.PP
.B estcmd merge [\-cl] db target
.PP
.B estcmd repair [\-rst|\-rsh] db
.PP
.B estcmd search [\-nl|\-nb] [\-pidx path] [\-ic enc] [\-vu|\-va|\-vf|\-vs|\-vh|\-vx|\-dd] [\-sn wnum hnum anum] [\-kn num] [\-um] [\-ec rn] [\-gs|\-gf|\-ga] [\-cd] [\-ni] [\-sf|\-sfr|\-sfu|\-sfi] [\-hs] [\-attr expr] [\-ord expr] [\-max num] [\-sk num] [\-aux num] [\-dis name] [\-sim id] db [phrase]
.PP
.B estcmd gather [\-tr] [\-cl] [\-ws] [\-no] [\-fe|\-ft|\-fh|\-fm] [\-fx sufs cmd] [\-fz] [\-fo] [\-rm sufs] [\-ic enc] [\-il lang] [\-bc] [\-lt num] [\-lf num] [\-pc enc] [\-px name] [\-aa name value] [\-apn|\-acc] [\-xs|\-xl|\-xh|\-xh2|\-xh3] [\-sv|\-si|\-sa] [\-ss name] [\-sd] [\-cm] [\-cs num] [\-ncm] [\-kn num] [\-um] db [file|dir]
.PP
.B estcmd purge [\-cl] [\-no] [\-fc] [\-pc enc] [\-attr expr] db [prefix]
.PP
.B estcmd extkeys [\-no] [\-fc] [\-dfdb file] [\-ncm] [\-ni] [\-kn num] [\-um] [\-attr expr] db [prefix]
.PP
.B estcmd words [\-nl|\-nb] [\-dfdb file] [\-kw|\-kt] db
.PP
.B estcmd draft [\-ft|\-fh|\-fm] [\-ic enc] [\-il lang] [\-bc] [\-lt num] [\-kn num] [\-um] [file]
.PP
.B estcmd break [\-ic enc] [\-il lang] [\-apn|\-acc] [\-wt] [file]
.PP
.B estcmd iconv [\-ic enc] [\-il lang] [\-oc enc] [file]
.PP
.B estcmd regex [\-inv] [\-repl str] expr [file]
.PP
.B estcmd scandir [\-tf|\-td] [\-pa|\-pu] [dir]
.PP
.B estcmd multi [\-db db] [\-nl|\-nb] [\-ic enc] [\-gs|\-gf|\-ga] [\-cd] [\-ni] [\-sf|\-sfr|\-sfu|\-sfi] [\-hs] [\-hu] [\-attr expr] [\-ord expr] [\-max num] [\-sk num] [\-aux num] [\-dis name] [phrase]
.PP
.B estcmd randput [\-ren|\-rla|\-reu|\-ror|\-rjp|\-rch] [\-cs num] db dnum
.PP
.B estcmd wicked db dnum
.PP
.B estcmd regression db
.PP
.B estcmd version

.SH DESCRIPTION
.PP
.B estcmd
is an aggregation of sub commands.  The name of a sub command is specified by the first argument.  Other arguments are parsed according to each sub command.  The argument
.I db
specifies the path of an index.
.TP
.B estcmd create [\-tr] [\-apn|\-acc] [\-xs|\-xl|\-xh|\-xh2|\-xh3] [\-sv|\-si|\-sa] [\-attr name type] db
Create an index.
.br
If
.B \-tr
is specified, a new index is created regardless if one exists.
.br
If
.B \-apn
is specified, N\-gram analysis is performed against European text also.
.br
If
.B \-acc
is specified, character category analysis is performed instead of N-gram analysis.
.br
If
.B \-xs
is specified, the index is tuned to register less than 50000 documents.
.br
If
.B \-xl
is specified, the index is tuned to register more than 300000 documents.
.br
If
.B \-xh
is specified, the index is tuned to register more than 1000000 documents.
.br
If
.B \-xh2
is specified, the index is tuned to register more than 5000000 documents.
.br
If
.B \-xh3
is specified, the index is tuned to register more than 10000000 documents.
.br
If
.B \-sv
is specified, scores are stored as void.
.br
If
.B \-si
is specified, scores are stored as 32-bit integer.
.br
If
.B \-sa
is specified, scores are stored as-is and marked not to be tuned when search.
.br
.B \-attr
specifies an attribute index and its data type.  This option can be specified multiple times.
.TP
.B estcmd put [\-tr] [\-cl] [\-apn|\-acc] [\-xs|\-xl|\-xh|\-xh2|\-xh3] [\-sv|\-si|\-sa] db [file]
Register a document of document draft to an index.
.br
.I file
specifies a target file.  If it is omitted, the standard input is read.
.br
If
.B \-tr
is specified, a new index is created regardless if one exists.
.br
If
.B \-cl
is specified, regions of a overwritten document are cleaned up.
.br
If
.B \-ws
is specified, scores are weighted statically with score weighting attribute.
.br
If
.B \-apn
is specified, N\-gram analysis is performed against European text also.
.br
If
.B \-acc
is specified, character category analysis is performed instead of N-gram analysis.
.br
If
.B \-xs
is specified, the index is tuned to register less than 50000 documents.
.br
If
.B \-xl
is specified, the index is tuned to register more than 300000 documents.
.br
If
.B \-xh
is specified, the index is tuned to register more than 1000000 documents.
.br
If
.B \-xh2
is specified, the index is tuned to register more than 5000000 documents.
.br
If
.B \-xh3
is specified, the index is tuned to register more than 10000000 documents.
.br
If
.B \-sv
is specified, scores are stored as void.
.br
If
.B \-si
is specified, scores are stored as 32-bit integer.
.br
If
.B \-sa
is specified, scores are stored as-is and marked not to be tuned when search.
.TP
.B estcmd out [\-pc enc] [\-cl] db expr
Remove information of a document from an index.
.br
.I expr
specifies the ID number, the URI, or the local path of a document.
.br
If
.B \-cl
is specified, regions of the document are cleaned up.
.br
.B \-pc
specifies the encoding of file paths.  By default, it is ISO-8859-1.
.TP
.B estcmd edit [\-pc enc] db expr name [value]
Edit an attribute of a document in an index.
.br
.I expr
specifies the ID number, the URI, or the local path of a document.
.br
.I name
specifies the name of an attribute.
.br
.I value
specifies the value of the attribute.  If it is omitted, the attribute is removed.
.br
.B \-pc
specifies the encoding of the file path and the attribute value.  By default, it is ISO-8859-1.
.TP
.B estcmd get [\-nl|\-nb] [\-pidx path] [\-pc enc] db expr [attr]
Output document draft of a document in an index.
.br
.I expr
specifies the ID number, the URI, or the local path of a document.
.br
If
.I attr
is specified, only the value of the attribute is output.
.br
If
.B \-nl
is specified, the index is opened without file locking.
.br
If
.B \-nb
is specified, file locking is performed without blocking.
.br
.B \-pidx
specifies the path of a pseudo index.  This option can be specified multiple times.
.br
.B \-pc
specifies the encoding of file paths.  By default, it is ISO-8859-1.
.TP
.B estcmd list [\-nl|\-nb] [\-lp] db
Output a list of all document in an index.
.br
If
.B \-nl
is specified, the index is opened without file locking.
.br
If
.B \-nb
is specified, file locking is performed without blocking.
.br
If
.B \-lp
is specified, local path equivalent to URL of "file://" is output.
.TP
.B estcmd uriid [\-nl|\-nb] [\-pidx path] [\-pc enc] db expr
Output the ID number of a document specified by URI.
.br
.I expr
specifies the URI or the local path of a document.
.br
If
.B \-nl
is specified, the index is opened without file locking.
.br
If
.B \-nb
is specified, file locking is performed without blocking.
.br
.B \-pidx
specifies the path of a pseudo index.  This option can be specified multiple times.
.br
.B \-pc
specifies the encoding of file paths.  By default, it is ISO-8859-1.
.TP
.B estcmd meta db [name [value]]
Handle meta data.
.br
.I name
specifies the name of a piece of meta data.  If it is omitted, a list of all names is output.
.br
.I value
specifies the value of the meta data to be recorded.  If it is omitted, the current value is output.  If it is an empty string, the meta data is removed.
.TP
.B estcmd inform [\-nl|\-nb] db
Output the number of documents and the number of unique words in an index.
.br
If
.B \-nl
is specified, the index is opened without file locking.
.br
If
.B \-nb
is specified, file locking is performed without blocking.
.TP
.B estcmd optimize [\-onp] [\-ond] db
Optimize an index and clean up dispensable regions.
.br
If
.B \-onp
is specified, it is omitted to clean up dispensable regions.
.br
If
.B \-ond
is specified, it is omitted to optimize the database files.
.TP
.B estcmd merge [\-cl] db target
Merge another index.
.br
.I target
specifies the path of another index.
.br
If
.B \-cl
is specified, regions of overwritten documents are cleaned up.
.TP
.B estcmd repair [\-rst|\-rsh] db
Repair a broken index.
.br
If
.B \-rst
is specified, strict consistency check is performed.
.br
If
.B \-rsh
is specified, consistency check is omitted.
.TP
.B estcmd search [\-nl|\-nb] [\-pidx path] [\-ic enc] [\-vu|\-va|\-vf|\-vs|\-vh|\-vx|\-dd] [\-sn wnum hnum anum] [\-kn num] [\-um] [\-ec rn] [\-gs|\-gf|\-ga] [\-cd] [\-ni] [\-sf|\-sfr|\-sfu|\-sfi] [\-hs] [\-attr expr] [\-ord expr] [\-max num] [\-sk num] [\-aux num] [\-dis name] [\-sim id] db [phrase]
Search an index for documents.
.br
.I phrase
specifies the search phrase.
.br
If
.B \-nl
is specified, the index is opened without file locking.
.br
If
.B \-nb
is specified, file locking is performed without blocking.
.br
.B \-pidx
specifies the path of a pseudo index.  This option can be specified multiple times.
.br
.B \-ic
specifies the input encoding.  By default, it is UTF\-8.
.br
If
.B \-vu
is specified, TSV of ID number and URI are output.
.br
If
.B \-va
is specified, multipart format including attributes is output.
.br
If
.B \-vf
is specified, multipart format including document draft is output.
.br
If
.B \-vs
is specified, multipart format including attributes and snippets is output.
.br
If
.B \-vh
is specified, human readable format including attributes and snippets is output.
.br
If
.B \-vx
is specified, XML including including attributes and snippets is output.
.br
If
.B \-dd
is specified, document draft data are dumped and saved into separated files.
.br
.B \-sn
specifies the number of whole width of snippet and width of strings picked up from the beginning of the text and width of strings picked up around each highlighted word.
.br
.B \-kn
specifies the number of keywords to be extracted.  By default, keyword extraction is not performed.
.br
If
.B \-um
is specified, morphological analyzers are used for keyword extraction.
.br
.B \-ec
specifies lower limit of similarity eclipse.
.br
If
.B \-gs
is specified, every key of N\-gram is checked.  By default, it is alternately.
.br
If
.B \-gf
is specified, keys of N\-gram are checked every three.
.br
If
.B \-ga
is specified, keys of N\-gram are checked every four.
.br
If
.B \-cd
is specified, whether documents match the search phrase definitely is checked.
.br
If
.B \-ni
is specified, TF\-IDF tuning is omitted.
.br
If
.B \-sf
is specified, the phrase is treated as a simplified form.
.br
If
.B \-sfr

oracc-hyperestraier/ChangeLog
2007-12-24  Mikio Hirabayashi  <mikio@fallabs.com>

	* mymorph.c (est_mecab_analyze): an option about EOS was added.

	- Release: 1.4.13

2007-11-18  Mikio Hirabayashi  <mikio@fallabs.com>

	* estraier.c: code clean-up.

	- Bugs related to type casting of perl binding were fixed.

	- Release: 1.4.12

2006-11-11  Mikio Hirabayashi  <mikio@fallabs.com>

	* estraier.c (est_resmap_add): "method" argument was added.

	* estraier.c (est_inode): a bug related to overflow was fixed.

	* estraier.c (est_narrow_score): a bug related to attribute distinction was fixed.

	* estraier.c (est_search_union): search accuracy with trailing uni-gram was improved.

	* estmaster.c (searchlocal): a bug related to scanning with regular expressions was fixed.

	* estscout.c (realmain): "union", "score", and "idsuffix" attributes were added.

	* estsupt.c (realmain): "score" attribute was added.

	* estsupt.c (procsearch): accuracy of hint was improved.

	* estsupt.c (realmain): "logmsg" parameter was added.

	* estscout.c (realmain): "distinct" parameter was added.

	* estscout.c (showresult): cache clear at random was added.

	- Release: 1.4.11

2006-11-11  Mikio Hirabayashi  <mikio@fallabs.com>

	* estraier.c (est_set_ecode): new function.

	* estraier.c (est_search_union): scoring of ASIS mode was modified.

	* estraier.c (est_resmap_add, est_resmap_dump_keys): new functions.

	* estraier.c (est_narrow_scores): efficiency of narrowing index was improved.

	* estraier.c (est_utime): new function.

	* estraier.c (est_cond_scores, est_cond_set_narrowing_scores): new functions.

	* estraier.c (est_rescc_put): a bug of memory leak was fixed.

	* estnode.c: the function "est_datum_printf" was replaced by "cbdatumprintf".

	* estmaster.c (sendnodecmdsearch): accuracy of hints was improved.

	* estfraud.c (sendnodecmdputdoc): accuracy of hints was improved.

	* estfraud.c (sendnodecmdputdoc): morphological analyzer support was added.

	* estfraud.c (sendnodecmdputdoc): accuracy of hints was improved.

	* estwaver.c (runinit, procinit): "-apn", "-acc", "-sv", "-si", "-sa" options was added.

	* estscout.c: new file.

	* estsupt.c: new file.

	- Release: 1.4.10

2006-11-03  Mikio Hirabayashi  <mikio@fallabs.com>

	* estraier.c (est_doc_score): new function.

	* estraier.c (est_db_check_option): new function.

	* estraier.c (est_search_union): a bug related to scoring when ASIS mode was fixed.

	* estraier.c (est_db_score_doc): difference operator was added.

	* estcmd.c (procinform): display of known options was added.

	* estcmd.c (doputdoc): modification date of files became able to be recorded as score.

	* estseek.c (showresult): accuracy of "{hnum}" operator was improved.

	* estwaver.c (runinit, procinit): "-xs", "-xl", "-xh" options were added.

	- "set_score" and "score" methods were added to language bindings and node APIs.

	- Release: 1.4.9

2006-11-03  Mikio Hirabayashi  <mikio@fallabs.com>

	* estfraud.c (realmain): handling path information was modified.

	* estseek.c (outputlog): "{hnum}" operator was added to the log format.

	- A bug related to optimization on Intel Mac series was escaped.

	- Release: 1.4.8

2006-10-20  Mikio Hirabayashi  <mikio@fallabs.com>

	* estraier.c (est_db_open): "ESTDBHUGE2" and "ESTDBHUGE3" options were added.

	* estraier.c (est_db_repair): a bug related to the attribute database was added.

	* estraier.c (est_cond_distinct): new function.

	* estmaster.c (startup): "extdelay" attribute was added.

	* estfraud.c (realmain): "extdelay" attribute was added.

	* estfraud.c (sendnodecmdsearch): a bug related to attribute search was fixed.

	* estseek.c (main): a bug related to FastCGI was fixed.

	- Release: 1.4.7

2006-10-15  Mikio Hirabayashi  <mikio@fallabs.com>

	* estraier.c (est_db_search_meta): a bug about the skip parameter was fixed.

	* estcmd.c (runmulti, procmulti): new functions.

	* estseek.c (showdoc): class name based on attribute name was added.

	* estseek.c (realmain): "helpfile" attribute was added.

	* estseek.c (shownaviform): new function.

	* estseek.c (realmain): "logformat" attribute was added.

	- A bug about URL encoding in the pure Ruby API was fixed.

	- Release: 1.4.6

2006-09-24  Mikio Hirabayashi  <mikio@fallabs.com>

	* estraier.c (est_search_aidx_attr): case insensitive mode was added.

	* estraier.c (est_check_cjk_only): new function.

	* estraier.c (est_make_cattr_list): search speed with CJK characters was improved.

	* estraier.c (est_db_add_attr_index): altering attribute index was added.

	* estraier.c (est_db_attr_index_exprs): new function.

	* estraier.c (est_regex_match_str): new function.

	* estraier.c (est_str_make_snippet): new function.

	* estraier.c (est_db_merge): a bug related to discordance of the ID attribute was fixed.

	* estraier.c (est_db_merge): division of inverted files was added.

	* estseek.c (showformforweb): new function.

	* estseek.c (realmain, showdoc): "deftitle" and "attrwidth" attributes were added.

	* estseek.c (showtitle): new function.

	* wavermod.c (waver_init, waver_open): "urlrule" attribute was added.

	* estproxy.c: new file.

	- Release: 1.4.5

2006-09-20  Mikio Hirabayashi  <mikio@fallabs.com>

	* estraier.c (est_db_pseudo_doc_num): new function.

	* estcmd.c (runregex, procregex): "-ci" option was added.


oracc-hyperestraier/rubynative/estcmd.rb
#! /usr/bin/ruby -w
#--
# Ruby binding of Hyper Estraier
#                                                       Copyright (C) 2004-2007 Mikio Hirabayashi
#  This file is part of Hyper Estraier.
#  Hyper Estraier is free software; you can redistribute it and/or modify it under the terms of
#  the GNU Lesser General Public License as published by the Free Software Foundation; either
#  version 2.1 of the License or any later version.  Hyper Estraier is distributed in the hope
#  that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Lesser General Public
#  License for more details.
#  You should have received a copy of the GNU Lesser General Public License along with Hyper
#  Estraier; if not, write to the Free Software Foundation, Inc., 59 Temple Place, Suite 330,
#  Boston, MA 02111-1307 USA.


$:.unshift("./src")
require "estraier"
include Estraier


# global constants
PROTVER = "1.0"
SEARCHMAX = 10
SEARCHAUX = 32
SNIPWWIDTH = 480
SNIPHWIDTH = 96
SNIPAWIDTH = 96
VM_ID = 0
VM_URI = 1
VM_ATTR = 2
VM_FULL = 3
VM_SNIP = 4


# main routine
def main(args)
  usage if args.length < 1
  rv = 0
  case args[0]
  when "put"
    rv = runput(args)
  when "out"
    rv = runout(args)
  when "edit"
    rv = runedit(args)
  when "get"
    rv = runget(args)
  when "uriid"
    rv = runuriid(args)
  when "inform"
    rv = runinform(args)
  when "optimize"
    rv = runoptimize(args)
  when "merge"
    rv = runmerge(args)
  when "search"
    rv = runsearch(args)
  else
    usage
  end
  return rv
end


# print usage and exit
def usage()
  STDERR.printf("%s: command line utility for the core API of Hyper Estraier\n", $0)
  STDERR.printf("\n")
  STDERR.printf("usage:\n")
  STDERR.printf("  %s put [-cl] [-ws] db [file]\n", $0)
  STDERR.printf("  %s out [-cl] db expr\n", $0)
  STDERR.printf("  %s edit db expr name [value]\n", $0)
  STDERR.printf("  %s get db expr [attr]\n", $0)
  STDERR.printf("  %s uriid db uri\n", $0)
  STDERR.printf("  %s inform db\n", $0)
  STDERR.printf("  %s optimize [-onp] [-ond] db\n", $0)
  STDERR.printf("  %s merge [-cl] db target\n", $0)
  STDERR.printf("  %s search [-vu|-va|-vf|-vs] [-gs|-gf|-ga] [-cd] [-ni] [-sf|-sfr|-sfu|-sfi]" \
                " [-attr expr] [-ord expr] [-max num] [-sk num] [-aux num] [-dis name]" \
                " db [phrase]\n", $0)
  STDERR.printf("\n")
  exit(1)
end


# print error string and flush the buffer */
def printerror(msg)
  STDERR.printf("%s: ERROR: %s\n", $0, msg)
  STDERR.flush
end


# parse arguments of the put command
def runput(args)
  dbname = nil
  file = nil
  opts = 0
  i = 1
  while i < args.length
    if !dbname && args[i] =~ /^-/
      if args[i] == "-cl"
        opts |= Database::PDCLEAN
      elsif args[i] == "-ws"
        opts |= Database::PDWEIGHT
      else
        usage
      end
    elsif !dbname
      dbname = args[i]
    elsif !file
      file = args[i]
    else
      usage
    end
    i += 1
  end
  usage if !dbname
  procput(dbname, file, opts)
end


# parse arguments of the out command
def runout(args)
  dbname = nil
  expr = nil
  opts = 0
  i = 1
  while i < args.length
    if !dbname && args[i] =~ /^-/
      if args[i] == "-cl"
        opts |= Database::ODCLEAN
      else
        usage
      end
    elsif !dbname
      dbname = args[i]
    elsif !expr
      expr = args[i]
    else
      usage
    end
    i += 1
  end
  usage if !dbname || !expr
  procout(dbname, expr, opts)
end


# parse arguments of the edit command
def runedit(args)
  dbname = nil
  expr = nil
  name = nil
  value = nil
  i = 1
  while i < args.length
    if !dbname && args[i] =~ /^-/
      usage
    elsif !dbname
      dbname = args[i]
    elsif !expr
      expr = args[i]
    elsif !name
      name = args[i]
    elsif !value
      value = args[i]
    else
      usage
    end
    i += 1
  end
  usage if !dbname || !expr || !name
  procedit(dbname, expr, name, value)
end


# parse arguments of the get command
def runget(args)
  dbname = nil
  expr = nil
  attr = nil
  i = 1
  while i < args.length
    if !dbname && args[i] =~ /^-/
      usage
    elsif !dbname
      dbname = args[i]
    elsif !expr
      expr = args[i]
    elsif !attr
      attr = args[i]
    else
      usage
    end
    i += 1
  end
  usage if !dbname || !expr
  procget(dbname, expr, attr)
end


# parse arguments of the uriid command
def runuriid(args)
  dbname = nil
  uri = nil
  i = 1
  while i < args.length
    if !dbname && args[i] =~ /^-/
      usage
    elsif !dbname
      dbname = args[i]
    elsif !uri
      uri = args[i]
    else
      usage
    end
    i += 1
  end
  usage if !dbname || !uri
  procuriid(dbname, uri)
end


# parse arguments of the inform command
def runinform(args)
  dbname = nil
  i = 1
  while i < args.length
    if !dbname && args[i] =~ /^-/
      usage
    elsif !dbname
      dbname = args[i]
    else
      usage
    end
    i += 1
  end
  usage if !dbname
  procinform(dbname)
end


# parse arguments of the optimize command
def runoptimize(args)
  dbname = nil
  opts = 0
  i = 1
  while i < args.length
    if !dbname && args[i] =~ /^-/
      if args[i] == "-onp"
        opts |= Database::OPTNOPURGE
      elsif args[i] == "-ond"
        opts |= Database::OPTNODBOPT
      else
        usage
      end
    elsif !dbname
      dbname = args[i]
    else
      usage
    end
    i += 1
  end
  usage if !dbname
  procoptimize(dbname, opts)
end


# parse arguments of the merge command
def runmerge(args)
  dbname = nil
  tgname = nil
  opts = 0
  i = 1
  while i < args.length
    if !dbname && args[i] =~ /^-/
      if args[i] == "-cl"
        opts |= Database::MGCLEAN
      else
        usage
      end
    elsif !dbname
      dbname = args[i]
    elsif !tgname
      tgname = args[i]
    else
      usage
    end
    i += 1
  end
  usage if !dbname || !tgname
  procmerge(dbname, tgname, opts)
end


# parse arguments of the search command
def runsearch(args)
  dbname = nil
  phrase = nil
  attrs = []
  ord = nil
  max = SEARCHMAX
  skip = 0
  opts = 0
  aux = SEARCHAUX
  dis = nil
  cd = false
  view = VM_ID
  i = 1
  while i < args.length
    if !dbname && args[i] =~ /^-/
      if args[i] == "-vu"
        view = VM_URI
      elsif args[i] == "-va"
        view = VM_ATTR
      elsif args[i] == "-vf"
        view = VM_FULL
      elsif args[i] == "-vs"
        view = VM_SNIP
      elsif args[i] == "-gs"
        opts |= Condition::SURE
      elsif args[i] == "-gf"
        opts |= Condition::FAST
      elsif args[i] == "-ga"
        opts |= Condition::AGITO
      elsif args[i] == "-cd"
        cd = true
      elsif args[i] == "-ni"
        opts |= Condition::NOIDF
      elsif args[i] == "-sf"
        opts |= Condition::SIMPLE
      elsif args[i] == "-sfr"
        opts |= Condition::ROUGH
      elsif args[i] == "-sfu"
        opts |= Condition::UNION
      elsif args[i] == "-sfi"
        opts |= Condition::ISECT
      elsif args[i] == "-attr"
        usage if (i += 1) >= args.length
        attrs.push(args[i])
      elsif args[i] == "-ord"
        usage if (i += 1) >= args.length
        ord = args[i]
      elsif args[i] == "-max"
        usage if (i += 1) >= args.length
        max = args[i].to_i
      elsif args[i] == "-sk"
        usage if (i += 1) >= args.length
        skip = args[i].to_i
      elsif args[i] == "-aux"
        usage if (i += 1) >= args.length
        aux = args[i].to_i
      elsif args[i] == "-dis"
        usage if (i += 1) >= args.length
        dis = args[i]
      else
        usage
      end
    elsif !dbname
      dbname = args[i]
    elsif !phrase
      phrase = args[i]
    else
      phrase += " " + args[i]
    end
    i += 1
  end
  usage if !dbname
  procsearch(dbname, phrase, attrs, ord, max, skip, opts, aux, dis, cd, view)
end


# perform the put command
def procput(dbname, file, opts)
  if file
    begin
      ifp = open(file, "rb")
      draft = ifp.read
    rescue
      printerror(file + ": could not open")
      return 1
    ensure
      ifp.close if ifp
    end
  else
    STDIN.binmode
    draft = STDIN.read
  end
  doc = Document::new(draft)
  db = Database::new
  unless db.open(dbname, Database::DBWRITER | Database::DBCREAT)
    printerror(dbname + ": " + db.err_msg(db.error))
    return 1
  end
  db.set_informer(Informer::new)
  unless db.put_doc(doc, opts)
    printerror(dbname + ": " + db.err_msg(db.error))
    db.close
    return 1
  end
  unless db.close
    printerror(dbname + ": " + db.err_msg(db.error))
    return 1
  end
  return 0
end


# perform the out command
def procout(dbname, expr, opts)
  db = Database::new
  unless db.open(dbname, Database::DBWRITER)
    printerror(dbname + ": " + db.err_msg(db.error))
    return 1
  end
  db.set_informer(Informer::new)
  id = expr.to_i
  if id < 1 && (id = db.uri_to_id(expr)) < 1
    printerror(dbname + ": " + db.err_msg(db.error))
    db.close
    return 1
  end
  unless db.out_doc(id, opts)
    printerror(dbname + ": " + db.err_msg(db.error))
    db.close
    return 1
  end
  unless db.close
    printerror(dbname + ": " + db.err_msg(db.error))
    return 1
  end
  return 0
end


# perform the edit command
def procedit(dbname, expr, name, value)
  db = Database::new
  unless db.open(dbname, Database::DBWRITER)
    printerror(dbname + ": " + db.err_msg(db.error))
    return 1
  end
  db.set_informer(Informer::new)
  id = expr.to_i
  if id < 1 && (id = db.uri_to_id(expr)) < 1
    printerror(dbname + ": " + db.err_msg(db.error))
    db.close
    return 1
  end
  unless doc = db.get_doc(id, Database::GDNOTEXT)
    printerror(dbname + ": " + db.err_msg(db.error))
    db.close
    return 1
  end
  doc.add_attr(name, value)
  unless db.edit_doc(doc)
    printerror(dbname + ": " + db.err_msg(db.error))
    db.close
    return 1
  end
  unless db.close
    printerror(dbname + ": " + db.err_msg(db.error))
    return 1
  end
  return 0
end


# perform the get command
def procget(dbname, expr, attr)
  db = Database::new
  unless db.open(dbname, Database::DBREADER)
    printerror(dbname + ": " + db.err_msg(db.error))
    return 1
  end
  id = expr.to_i
  if id < 1 && (id = db.uri_to_id(expr)) < 1
    printerror(dbname + ": " + db.err_msg(db.error))
    db.close
    return 1
  end
  if attr
    unless value = db.get_doc_attr(id, attr)
      printerror(dbname + ": " + db.err_msg(db.error))
      db.close
      return 1
    end
    printf("%s\n", value)
  else
    unless doc = db.get_doc(id, 0)
      printerror(dbname + ": " + db.err_msg(db.error))
      db.close
      return 1
    end
    printf("%s", doc.dump_draft)
  end
  unless db.close
    printerror(dbname + ": " + db.err_msg(db.error))
    return 1
  end
  return 0
end


# perform the uriid command
def procuriid(dbname, uri)
  db = Database::new
  unless db.open(dbname, Database::DBREADER)
    printerror(dbname + ": " + db.err_msg(db.error))
    return 1
  end
  unless (id = db.uri_to_id(uri)) > 0
    printerror(dbname + ": " + db.err_msg(db.error))
    db.close
    return 1
  end
  printf("%d\n", id)
  unless db.close
    printerror(dbname + ": " + db.err_msg(db.error))
    return 1
  end
  return 0
end


# perform the inform command
def procinform(dbname)
  db = Database::new
  unless db.open(dbname, Database::DBREADER)
    printerror(dbname + ": " + db.err_msg(db.error))
    return 1
  end
  printf("number of documents: %d\n", db.doc_num)
  printf("number of words: %d\n", db.word_num)
  printf("file size: %d\n",  db.size.to_i)
  unless db.close
    printerror(dbname + ": " + db.err_msg(db.error))
    return 1
  end
  return 0
end


# perform the optimize command
def procoptimize(dbname, opts)
  db = Database::new
  unless db.open(dbname, Database::DBWRITER)
    printerror(dbname + ": " + db.err_msg(db.error))
    return 1
  end
  db.set_informer(Informer::new)
  unless db.optimize(opts)
    printerror(dbname + ": " + db.err_msg(db.error))
    db.close
    return 1
  end
  unless db.close
    printerror(dbname + ": " + db.err_msg(db.error))
    return 1
  end
  return 0
end


# perform the merge command
def procmerge(dbname, tgname, opts)
  db = Database::new
  unless db.open(dbname, Database::DBWRITER)
    printerror(dbname + ": " + db.err_msg(db.error))
    return 1
  end
  db.set_informer(Informer::new)
  unless db.merge(tgname, opts)
    printerror(dbname + ": " + db.err_msg(db.error))
    db.close
    return 1
  end
  unless db.close
    printerror(dbname + ": " + db.err_msg(db.error))
    return 1
  end
  return 0
end


# perform the search command
def procsearch(dbname, phrase, attrs, ord, max, skip, opts, aux, dis, cd, view)
  db = Database::new
  unless db.open(dbname, Database::DBREADER)
    printerror(dbname + ": " + db.err_msg(db.error))
    return 1
  end
  cond = Condition::new()
  cond.set_phrase(phrase) if phrase
  for i in 0...attrs.length
    cond.add_attr(attrs[i])
  end
  cond.set_order(ord) if ord
  cond.set_max(max) if max >= 0
  cond.set_skip(skip) if skip >= 0
  cond.set_options(opts)
  cond.set_auxiliary(aux)
  cond.set_distinct(dis) if dis
  stime = Time::now.to_f
  res = db.search(cond)
  etime = Time::now.to_f
  border = "--------[" + Time::now.to_f.to_s.gsub(/\./, "") + "]--------"
  printf("%s\n", border)
  printf("VERSION\t%s\n", PROTVER)
  printf("NODE\tlocal\n")
  printf("HIT\t%d\n", res.hint(""))
  snwords = []
  words = res.hint_words
  for i in 0...words.length
    word = words[i]
    hits = res.hint(word)
    snwords.push(word) if hits > 0
    printf("HINT#%d\t%s\t%d\n", i + 1, word, hits)
  end
  printf("TIME\t%0.3f\n", etime - stime)
  printf("DOCNUM\t%d\n", db.doc_num())
  printf("WORDNUM\t%d\n", db.word_num())
  if view == VM_URI
    printf("VIEW\tURI\n")
  elsif view == VM_ATTR
    printf("VIEW\tATTRIBUTE\n")
  elsif view == VM_FULL
    printf("VIEW\tFULL\n")
  elsif view == VM_SNIP
    printf("VIEW\tSNIPPET\n")
  else
    printf("VIEW\tID\n")
  end
  printf("\n")
  printf("%s\n", border) if view != VM_ATTR && view != VM_FULL && view != VM_SNIP
  dnum = res.doc_num()
  for i in 0...dnum
    id = res.get_doc_id(i)
    if view == VM_URI
      next unless doc = db.get_doc(id, cd ? 0 : Database::GDNOTEXT)
      next if cd && !db.scan_doc(doc, cond)
      printf("%d\t%s\n", id, doc.attr("@uri"))
    elsif view == VM_ATTR
      next unless doc = db.get_doc(id, cd ? 0 : Database::GDNOTEXT)
      next if cd && !db.scan_doc(doc, cond)
      printf("%s\n", border)
      names = doc.attr_names()
      for j in 0...names.length
        printf("%s=%s\n", names[j], doc.attr(names[j]))
      end
      printf("\n")
    elsif view == VM_FULL
      next unless doc = db.get_doc(id, 0)
      next if cd && !db.scan_doc(doc, cond)
      printf("%s\n", border)
      printf("%s", doc.dump_draft())
    elsif view == VM_SNIP
      next unless doc = db.get_doc(id, 0)
      next if cd && !db.scan_doc(doc, cond)
      printf("%s\n", border)
      names = doc.attr_names()
      for j in 0...names.length
        printf("%s=%s\n", names[j], doc.attr(names[j]))
      end
      printf("\n")
      printf("%s", doc.make_snippet(snwords, SNIPWWIDTH, SNIPHWIDTH, SNIPAWIDTH))
    else
      printf("%d\n", id)
    end
  end
  printf("%s:END\n", border)
  unless db.close
    printerror(dbname + ": " + db.err_msg(db.error))
    return 1
  end
  return 0
end


# class for callback function for database events
class Informer
  def inform(message)
    printf("%s: INFO: %s\n", $0, message)
  end
end


# perform the main routine
$0.gsub!(/.*\//, "")
exit(main(ARGV))



# END OF FILE
